TurboMinerVirtualTable.Unity/Assets/Scripts/Networking/Server.cs

```diff
@@ -0,0 +1,251 @@
+        clients = new List<ServerClient>();
+        disconnectList = new List<ServerClient>();
+
+        try
+        {
+            server = new TcpListener(IPAddress.Any, Port);
+            server.Start();
+
+            StartListening();
+            serverStarted = true;
+
+            readThread = new Thread(new ThreadStart(ReadData));
+            readThread.IsBackground = true;
+            readThread.Start();
+        }
+        catch (Exception e)
+        {
+            Debug.Log($"Socket error: {e.Message}");
+        }
+    }
+
+    private void ReadData()
+    {
+        while (!threadStop)
+        {
+            if (!serverStarted)
+            {
+                continue;
+            }
+
+            foreach (var client in clients.ToList()) //ToList because list may change during foreach loop, so make copy to prevent that
+            {
+                if (client == null)
+                {
+                    continue;
+                }
+
+                if (!IsConnected(client.Tcp))
+                {
+                    client.Tcp.Close();
+                    disconnectList.Add(client);
+                    continue;
+                }
+
+                var networkStream = client.Tcp.GetStream();
+
+                if (networkStream.DataAvailable)
+                {
+                    var streamReader = new StreamReader(networkStream, true);
+                    var data = streamReader.ReadLine();
+                    if (data != null)
+                    {
+                        OnIncomingData(client, data);
+                    }
+                }
+            }
+
+            for (int i = 0; i < disconnectList.Count - 1; ++i)
+            {
+                // Tell our player somebody has disconnected
+                clients.Remove(disconnectList[i]);
+                disconnectList.RemoveAt(i);
+            }
+        }
+    }
+
+    private void StartListening()
+    {
+        server.BeginAcceptTcpClient(AcceptTcpClient, server);
+    }
+
+    private void AcceptTcpClient(IAsyncResult result)
+    {
+        var listener = (TcpListener)result.AsyncState;
+
+        // get all clients names to message
+        var allUseres = string.Empty;
+        foreach (var client in clients)
+        {
+            allUseres += $"{client.ClientName}|";
+        }
+
+        var serverClient = new ServerClient(listener.EndAcceptTcpClient(result));
+        clients.Add(serverClient);
+        Debug.Log("Somebody has connected!");
+
+        StartListening();
+
+        Broadcast($"{MessageCommands.Server.Who}|{allUseres}", serverClient);
+    }
+
+    private bool IsConnected(TcpClient client)
+    {
+        try
+        {
+            if (client != null && client.Client != null && client.Connected)
+            {
+                if (client.Client.Poll(0, SelectMode.SelectRead))
+                {
+                    return client.Client.Receive(new byte[1], SocketFlags.Peek) != 0;
```

Comment:
client.Client->client(client) :co:

Reply:
no to nazewnictwo zmienie (btw ten kod to jest jedyna czesc ktora skopiowalem bez zrozumienia od takiego typa, ktory tez powiedzial ze on tego nie rozumie xDDD) ale ponoc to costam optymalizuje jak ktos sie rozlaczy, trzeba by zresearchowac ten temat